api/fake_agent_server.go

```diff
@@ -0,0 +1,176 @@
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		panic(fmt.Sprintf("fake: failed to encode register stack response: %v", err))
+	}
+}
+
+func (f *FakeAgentServer) handleStacksAPI(w http.ResponseWriter, r *http.Request) {
+	// Route based on path
+	if r.URL.Path == "/stacks/test-stack/scheduled-jobs/batch-reserve" {
+		f.handleReserveJobs(w, r)
+		return
+	}
+
+	// Default: not found
+	http.Error(w, "not found: "+r.URL.Path, http.StatusNotFound)
+}
+
+func (f *FakeAgentServer) handleReserveJobs(w http.ResponseWriter, r *http.Request) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if r.Method != http.MethodPut && r.Method != http.MethodPost {
+		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	var req stacksapi.BatchReserveJobsRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// Record the call
+	jobIDsCopy := make([]string, len(req.JobUUIDs))
+	copy(jobIDsCopy, req.JobUUIDs)
+	f.ReserveCalls = append(f.ReserveCalls, jobIDsCopy)
+
+	// Return configured error if set
+	if f.ReserveError != "" {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(f.ReserveStatusCode)
+		if err := json.NewEncoder(w).Encode(map[string]string{
+			"message": f.ReserveError,
+		}); err != nil {
+			panic(fmt.Sprintf("fake: failed to encode error response: %v", err))
+		}
+		return
+	}
+
+	// Return configured response
+	var resp stacksapi.BatchReserveJobsResponse
+	if f.ReserveResponse != nil {
+		resp = *f.ReserveResponse
+	} else {
+		// Default: all jobs reserved successfully
+		resp = stacksapi.BatchReserveJobsResponse{
+			Reserved:    jobIDsCopy,
+			NotReserved: []string{},
+		}
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(f.ReserveStatusCode)
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		panic(fmt.Sprintf("fake: failed to encode reserve jobs response: %v", err))
+	}
+}
+
+// ReserveCallCount returns the number of times ReserveJobs was called.
+func (f *FakeAgentServer) ReserveCallCount() int {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	return len(f.ReserveCalls)
+}
+
+// GetReserveCallIDs returns the job IDs from a specific ReserveJobs call.
+func (f *FakeAgentServer) GetReserveCallIDs(callIndex int) ([]string, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if callIndex < 0 || callIndex >= len(f.ReserveCalls) {
+		return nil, fmt.Errorf("call index %d out of range (0-%d)", callIndex, len(f.ReserveCalls)-1)
+	}
+	ids := make([]string, len(f.ReserveCalls[callIndex]))
+	copy(ids, f.ReserveCalls[callIndex])
+	return ids, nil
+}
+
+// GetReserveCalls returns a copy of all recorded ReserveJobs calls.
+func (f *FakeAgentServer) GetReserveCalls() [][]string {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	calls := make([][]string, len(f.ReserveCalls))
+	for i := range f.ReserveCalls {
+		calls[i] = make([]string, len(f.ReserveCalls[i]))
+		copy(calls[i], f.ReserveCalls[i])
+	}
+	return calls
```

Comment:
I'm not convinced these methods are worthwhile. If you are only checking them after finishing the client call, then they shouldn't be concurrent with any API handling. (You could call `Close` to be sure.) If they are meant to be concurrent with API handling then the return values are unreliable.